servicetalk-transport-api/src/main/java/io/servicetalk/transport/api/EarlyConnectionAcceptor.java

```diff
@@ -0,0 +1,51 @@
+/*
+ * Copyright © 2023 Apple Inc. and the ServiceTalk project authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.servicetalk.transport.api;
+
+import io.servicetalk.concurrent.api.AsyncCloseable;
+import io.servicetalk.concurrent.api.Completable;
+
+import static io.servicetalk.concurrent.api.Completable.completed;
+
+/**
+ * Allows to accept or reject connections early in the connection setup stage.
+ */
+@FunctionalInterface
+public interface EarlyConnectionAcceptor extends ExecutionStrategyInfluencer<ConnectExecutionStrategy>, AsyncCloseable {
+
+    /**
+     * Accept or reject an incoming connection.
+     *
+     * @param info additional information about the connection to make an acceptance decision.
+     * @return a completed (to accept) or a failed (to reject) {@link Completable}
+     */
+    Completable accept(ReducedConnectionInfo info);
```

Comment:
Let's assume that we change this type to `ConnectionInfo`. Then `EarlyConnectionAcceptor` will be identical to `LateConnectionAcceptor`. We will be able to reuse the same interface keeping different configuration methods on the server builder. But then I'm not sure what will be a good name for a new interface. Doesn't look like we can reuse the existing `ConnectionAcceptor`. WDYT?

Reply:
Agreed that a common interface would be ok, but can't figure out a good name for it since we cannot reuse the current `ConnectionAcceptor` yeah. On the other hand, having 2 interfaces might not be too bad since it would allow to evolve them differently in the future if needed and since most people will use it with a lambda anyways it will not be as visible on the user-facing APIs.

So In my opinion if we do not come up with a good name for a common interface, I would be okay with keeping them separate.

Reply:
Good points!
Also, with a single interface it will be possible to reuse the same impl in early and later stage, which is likely more error prone. For example, users may end up not having `SSLSession` and rejecting requests unintentionally if the place "later impl" in early stage, or they may consume more resources by placing "early impl" at later stage. So, too easy to misuse. Better to force people to think about the position and behavior when they implement interface instead of when they find a good spot for existing impl.